src/LuaUtils.cpp

```diff
@@ -171,127 +171,363 @@ static bool _import_core(lua_State *L, const std::string &importName)
+
+		// check original name
+		isImported = get_cached(L, cacheName);
+
+		// check name with extension
+		if (!isImported && !importInfo.isFullName)
+		{
+			cacheName += ".lua";
+			isImported = get_cached(L, cacheName);
+		}
+
+		if (isImported)
+		{
+			DEBUG_PRINTF(" found cached %s\n", cacheName.c_str());
+			break;
+		}
 	}
-	lua_pop(L, 1);
-	DEBUG_PRINTF("import: not cached %s\n", cacheName.c_str());
 
-	// next we're going to look in libs/
-	cacheName = FileSystem::JoinPath("libs", importName+".lua");
-	lua_getfield(L, -1, cacheName.c_str());
-	if (!lua_isnil(L, -1)) {
-		// found it, use it
-		DEBUG_PRINTF("import: found cached %s\n", cacheName.c_str());
-		lua_remove(L, -2);
-		return 1;
+	if (!isImported)
+		DEBUG_PRINTF(" not cached\n");
+
+	LUA_DEBUG_END(L, (isImported ? 1 : 0));
+	return isImported;
+}
+
+/**
+* Tries to load file.
+* Returns true if file exists and puts return of dofile to stack.
+*/
+static bool load_file(lua_State *L, const std::string& name)
+{
+	LUA_DEBUG_START(L);
+
+	bool loaded = false;
+	RefCountedPtr<FileSystem::FileData> fileData = FileSystem::gameDataFiles.ReadFile(name);
+
+	// if file exists
+	if (fileData)
+	{
+		DEBUG_PRINTF(" loading %s...\n", name.c_str());
+		pi_lua_dofile(L, *fileData, 1);
+		loaded = true;
 	}
-	lua_pop(L, 1);
-	DEBUG_PRINTF("import: not cached %s\n", cacheName.c_str());
 
-	RefCountedPtr<FileSystem::FileData> code = FileSystem::gameDataFiles.ReadFile(cacheName);
+	LUA_DEBUG_END(L, (loaded ? 1 : 0));
+	return loaded;
+}
 
-	// if its not found then we try to find the lib relative to the calling file
-	if (!code) {
-		DEBUG_PRINTF("import: not found %s, trying relative\n", cacheName.c_str());
-
-		lua_Debug ar;
-		lua_getstack(L, 1, &ar);
-		lua_getinfo(L, "S", &ar);
-
-		if (ar.source) {
-			int start = 0;
-			int end = strlen(ar.source);
-
-			if (ar.source[start] != '@' && end > 5)
-				start = 4;
-
-			if (ar.source[start] == '@') {
-				start++;
-
-				static char source[1024];
-				strncpy(source, &ar.source[start], end-start+1);
-				DEBUG_PRINTF("import: called from %s\n", source);
-
-				for (end = end-start-1; end >= 0; end--)
-					if (source[end] == '/') {
-						source[end] = '\0';
-						break;
-					}
-
-				DEBUG_PRINTF("import: relative path %s\n", source);
-
-				cacheName = FileSystem::JoinPath(source, importName+".lua");
-				lua_getfield(L, -1, cacheName.c_str());
-				if (!lua_isnil(L, -1)) {
-					// found it, use it
-					DEBUG_PRINTF("import: found cached %s\n", cacheName.c_str());
-					lua_remove(L, -2);
-					return 1;
-				}
-				lua_pop(L, 1);
-				DEBUG_PRINTF("import: not cached %s\n", cacheName.c_str());
-
-				code = FileSystem::gameDataFiles.ReadFile(cacheName);
-			}
+/**
+* Tries to load file.
+* If file found returns true, puts return of dofile to stack, no matter nil it or value.
+* If import was successful, changes fileName in the importInfo to the filename that was found.
```

Comment:
Hum. The doc says it changes something in the importInfo struct while the function signature uses a const ref. Which is it?

Reply:
@laarmen it means I cant change importInfo object in full, but can change values. I thought about it early and decided to make consts in the struct:
~~~cpp
// simple struct to pass data between functions
struct ImportInfo
{
	const std::string& importName; // original name argument from "import(importName)"
	std::string& fileName; // name of the existing file
	const bool& isFullName; // if true import will no try to load relative to the importDirectories
	std::vector<std::string>& importDirectories; // contans list of paths where file can be imported
};
~~~
Maybe it should be changed, but I dont know how to make it elegant.

Reply:
I don't like, and I'm actually surprised it compiles.
Please drop the const qualifier for the ref, it is IMO misleading.